### PR TITLE
Display license content

### PR DESCRIPTION
**Pull request by zhhyu**
Display license content when the path is good.
Fix: https://github.com/NuGet/Engineering/issues/1852
Fix: https://github.com/NuGet/Engineering/issues/1851

### CODE REVIEW

File: src/NuGet.Licenses/Services/LicenseFileService.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+namespace NuGet.Licenses
+{
+    public class LicenseFileService : ILicenseFileService
+    {
+        public string GetLicenseFilePath(string licenseIdentifier)
+        {
+            return Path.GetFullPath(Path.Combine(HttpContext.Current.Server.MapPath("~\\App_Data\\Licenses\\"), String.Concat(licenseIdentifier, ".txt")));
+        }
+
+        public bool IsLicenseFilePathAllowed(string licenseFilePath)
```
Comment (agr) [inline src/NuGet.Licenses/Services/LicenseFileService.cs]: I'd suggest to remove this method and instead always check for path correctness in `GetLicenseFilePath` and throw an exception if it's not correct. Then calling code would not have a chance not to check for the path correctness.

File: src/NuGet.Licenses/Services/LicenseFileService.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+using System;
+using System.IO;
+using System.Web;
+namespace NuGet.Licenses
+{
+    public class LicenseFileService : ILicenseFileService
+    {
+        public string GetLicenseFilePath(string licenseIdentifier)
+        {
+            return Path.GetFullPath(Path.Combine(HttpContext.Current.Server.MapPath("~\\App_Data\\Licenses\\"), String.Concat(licenseIdentifier, ".txt")));
```
Comment (agr) [inline src/NuGet.Licenses/Services/LicenseFileService.cs]: We'll have to test this code and in its current form it is quite untestable. I suggest wrapping all file I/O calls into their own method, so we'd have something like:
```C#
interface IFileService
{
    string GetFullPath(string relativePath);
    bool FileExists(string path);
    string ReadAllText(string path);
}
```
And separately, move `HttpContext` operations into it's own object, so for tests we wouldn't have to construct HttpContext, but mock that object instead.

File: src/NuGet.Licenses/Services/LicenseFileService.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        }
+
+        public bool IsLicenseFilePathAllowed(string licenseFilePath)
+        {
+            return licenseFilePath.StartsWith(HttpContext.Current.Server.MapPath("~\\App_Data\\Licenses\\"));
+        }
+
+        public bool IsLicenseFileExisted(string licenseFilePath)
+        {
```
Comment (agr) [inline src/NuGet.Licenses/Services/LicenseFileService.cs]: We need to check `IsLicenseFilePathAllowed` here before checking if file exists.

File: src/NuGet.Licenses/Services/LicenseFileService.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+            return File.Exists(licenseFilePath);
+        }
+
+        public string GetLicenseFileContent(string licenseFilePath)
+        {
```
Comment (agr) [inline src/NuGet.Licenses/Services/LicenseFileService.cs]: And here as well.

File: src/NuGet.Licenses/Views/License/DisplayLicense.cshtml [hunk 5]
```diff
@@ -5,4 +5,10 @@
 
 <h2>@Model.Title</h2>
 
-@Model.Text
\ No newline at end of file
+<pre>@Model.Text </pre>
+<style>
```
Comment (agr) [inline src/NuGet.Licenses/Views/License/DisplayLicense.cshtml]: Please remove the inline style. The target branch now has `Site.css` you can update it instead.

File: src/NuGet.Licenses/Services/ILicenseFileService.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        /// The function is used to check whether the license file path is allowed.
+        /// </summary>
+        /// <param name="licenseFilePath">The license file path</param>
+        bool IsLicenseFilePathAllowed(string licenseFilePath);
+
+        /// <summary>
+        /// The function is used to check whether the license file path is existed.
+        /// </summary>
+        /// <param name="licenseFilePath">The license file path</param>
+        bool IsLicenseFileExisted(string licenseFilePath);
```
Comment (agr) [inline src/NuGet.Licenses/Services/ILicenseFileService.cs]: `IsLicenseFilePresent`? `DoesLicenseFileExist`? Current name is wrong.

**Comment by agr** [comment]
Also, please add the script that generates license files.

File: src/NuGet.Licenses/Services/LicenseFileService.cs (hunk 1 above)
Comment (zhhyu) [reply src/NuGet.Licenses/Services/LicenseFileService.cs]: Done

File: src/NuGet.Licenses/Services/LicenseFileService.cs (hunk 2 above)
Comment (zhhyu) [reply src/NuGet.Licenses/Services/LicenseFileService.cs]: Create a new service to get folder path.

File: src/NuGet.Licenses/Services/LicenseFileService.cs (hunk 3 above)
Comment (zhhyu) [reply src/NuGet.Licenses/Services/LicenseFileService.cs]: Done

File: src/NuGet.Licenses/Services/LicenseFileService.cs (hunk 4 above)
Comment (zhhyu) [reply src/NuGet.Licenses/Services/LicenseFileService.cs]: Add check.

File: src/NuGet.Licenses/Views/License/DisplayLicense.cshtml (hunk 5 above)
Comment (zhhyu) [reply src/NuGet.Licenses/Views/License/DisplayLicense.cshtml]: Delete it

File: src/NuGet.Licenses/Services/ILicenseFileService.cs (hunk 6 above)
Comment (zhhyu) [reply src/NuGet.Licenses/Services/ILicenseFileService.cs]: Change to another name.

File: src/NuGet.Licenses/Services/ILicensesFolderPathService.cs [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+{
+    /// <summary>
+    /// This interface is used to get the path of the licenses folder.
+    /// /// </summary>
+    public interface ILicensesFolderPathService
+    {
+        /// <summary>
+        /// The function is used to get the path of the licenses folder.
+        /// </summary>
+        string getLicensesFolderPath();
```
Comment (agr) [inline src/NuGet.Licenses/Services/ILicensesFolderPathService.cs]: First character should be in upper case.

File: src/NuGet.Licenses/Global.asax.cs [hunk 8]
```diff
@@ -86,6 +86,13 @@ private static void SetupDependencyInjection()
                 .As<ILicenseFileService>()
                 .SingleInstance();
 
+            builder
+                .RegisterType<LicensesFolderPathService>()
+                .AsSelf()
```
Comment (agr) [inline src/NuGet.Licenses/Global.asax.cs]: We shouldn't be adding classes as self.

File: src/NuGet.Licenses/Global.asax.cs [hunk 9]
```diff
@@ -86,6 +86,13 @@ private static void SetupDependencyInjection()
...
+                .As<ILicensesFolderPathService>()
+                .SingleInstance();
```
Comment (agr) [inline src/NuGet.Licenses/Global.asax.cs]: `InstancePerLifetimeScope`

File: src/NuGet.Licenses/Global.asax.cs [hunk 10]
```diff
@@ -86,6 +86,13 @@ private static void SetupDependencyInjection()
                 .As<ILicenseFileService>()
                 .SingleInstance();
```
Comment (agr) [inline src/NuGet.Licenses/Global.asax.cs]: `InstancePerLifetimeScope`

File: src/NuGet.Licenses/Services/LicenseFileService.cs (hunk 2 above)
Comment (agr) [reply src/NuGet.Licenses/Services/LicenseFileService.cs]: Need file operations abstractions as well.

File: src/NuGet.Licenses/Services/ILicensesFolderPathService.cs (hunk 7 above)
Comment (zhhyu) [reply src/NuGet.Licenses/Services/ILicensesFolderPathService.cs]: Done

**Pull request merged**